### PR TITLE
Fixed widget IDs to be reusable after window closes

### LINKED ISSUE #2514: Widget IDs are not cleared when a window closes
### Describe the bug

When a window is created upon the click of a button or other widget in a main window, IDs of widgets within that sub-window are not cleared, causing window creation and displaying to fail when the window is closed and re-opened.

### Steps to reproduce

1. Create any main window with some widget that instantiates and shows another window when the widget is pressed, clicked, etc.
2. The sub-window should use widget IDs, such as paths (I use this format for determining which widget to update when a network event is received in my application).
3. Close the window and try to re-open it. You will get an error along the lines of `KeyError: "There is already a widget with the id '/devices/0/inputs/0/sends/0/Gain/value'"`

### Expected behavior

Widget IDs should be cleared when the window is either closed, deleted, or otherwise ceases to exist.

### Screenshots

_No response_

### Environment

- Operating System: Windows 10 22H2 (AMD64) build 19045.4291
- Python version: 3.12.2
- Software versions:
  - Briefcase: 0.3.17
  - Toga: 0.4.2+

### Logs

```
[19:49:53] Error in handler: "There is already a widget with the id '/devices/0/inputs/0/sends/0/Gain/value'"                                                      subprocess.py:681
           Traceback (most recent call last):                                                                                                                      subprocess.py:681
             File "C:\Users\ethin\source\venvs\uaccess\Lib\site-packages\toga\handlers.py", line 80, in _handler                                                   subprocess.py:681
               result = handler(interface, *args, **kwargs)                                                                                                        subprocess.py:681
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                        subprocess.py:681
             File "C:\Users\ethin\source\uaccess\src\uaccess\app.py", line 216, in open_sends                                                                      subprocess.py:681
               self.sends_dialogs[widget.id].build()                                                                                                               subprocess.py:681
             File "C:\Users\ethin\source\uaccess\src\uaccess\dialogs\sends_dialog.py", line 27, in build                                                           subprocess.py:681
               self.sends_content.add(edit)                                                                                                                        subprocess.py:681
             File "C:\Users\ethin\source\venvs\uaccess\Lib\site-packages\toga\widgets\base.py", line 100, in add                                                   subprocess.py:681
               child.window = self.window                                                                                                                          subprocess.py:681
               ^^^^^^^^^^^^                                                                                                                                        subprocess.py:681
             File "C:\Users\ethin\source\venvs\uaccess\Lib\site-packages\toga\widgets\base.py", line 227, in window                                                subprocess.py:681
               window.app.widgets._add(self)                                                                                                                       subprocess.py:681
             File "C:\Users\ethin\source\venvs\uaccess\Lib\site-packages\toga\app.py", line 192, in _add                                                           subprocess.py:681
               raise KeyError(f"There is already a widget with the id {widget.id!r}")                                                                              subprocess.py:681
           KeyError: "There is already a widget with the id '/devices/0/inputs/0/sends/0/Gain/value'"                                                              subprocess.py:681
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by proneon267**
Fixes #2514 

Fixed widget IDs to be reusable after window closes. Also added a new test to check the behavior.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

File: core/src/toga/window.py [hunk 1]
```diff
@@ -272,6 +272,9 @@ def close(self) -> None:
         undefined, except for :attr:`closed` which can be used to check if the window
         was closed.
         """
+        if getattr(self, "content", None) is not None:
```
Comment (freakboy3742) [inline core/src/toga/window.py:275]: In what conditions can content *not* exist as an attribute? It's one of the first attributes set on Window

File: core/src/toga/window.py [hunk 2]
```diff
@@ -272,6 +272,9 @@ def close(self) -> None:
...
+            self.content.clear()
```
Comment (freakboy3742) [inline core/src/toga/window.py:276]: Why is this needed? AFAICT, setting the window to None should be sufficient, and will propagate to all children. Clearing the children from content means the ownership hierarchy of `content` will be destroyed.

File: core/tests/test_window.py [hunk 3]
```diff
@@ -381,6 +381,49 @@ def test_screen_position(window, app):
     assert window.screen_position == (100, 100)
 
 
+def test_widget_id_reusable_after_window_closes(window, app):
+    """Widget IDs can be reused after the associated widget's window is closed."""
+
+    # Common widget IDs
+    common_widget_ids = {"content": "window_content", "label": "sample_label"}
```
Comment (freakboy3742) [inline core/tests/test_window.py:388]: I don't see why this is a dictionary containing 2 constants, instead of just using 2 constants. The "dictionary" features of this are completely unused - we just need 2 constant strings.

File: core/tests/test_window.py [hunk 4]
```diff
@@ -381,6 +381,49 @@ def test_screen_position(window, app):
...
+
+    # Create a second window and create its content and children & specify their IDs.
+    second_window = toga.Window()
+    second_window.content = toga.Box(
+        id=common_widget_ids["content"],
+        children=[toga.Label(text="Sample Label", id=common_widget_ids["label"])],
+    )
+    # Show the second window and check that it is in the app's windows registry.
```
Comment (freakboy3742) [inline core/tests/test_window.py:396]: There seems to be some confusion here. We aren't concerned about the *window* registry at all (not that there even is one. There's a list of windows). 

What we care about is the *widget* registry. The tests you've got here verify the *window* has been created, but that's not the thing that is the source of the bugs here.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I believe you've correctly identified the source of the bug here; but both the fix and the test miss a lot of details.

File: core/src/toga/window.py (hunk 1 above)
Comment (proneon267) [reply core/src/toga/window.py]: sorry, I meant to do:
```
if self.content:
```

File: core/src/toga/window.py (hunk 2 above)
Comment (proneon267) [reply core/src/toga/window.py]: Thanks, I hadn't noticed `_remove()` also propagates to the children as well. Also, does ownership hierarchy matter when we are clearing all the widgets?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This is a lot closer. The implementation is simpler, and the tests are actually validating the thing that has been changed. 

However, the tests are only verifying the "good" path through the code - there's no validation of the "bad" path. Consider the cases of:

1. Adding the a widget with a re-used ID to an existing (and visible) window
2. Constructing a new window that re-uses an ID

It should be legal to *create* a widget with a duplicated ID; it's only when the widget is added to a window that a problem will occur (as widgets don't appear in the registry until they're part of a layout). 

There's also a difference between a parent widget with a duplicated ID and a *child* widget with a duplicated ID. Both will raise an error - but the paths to generating that error may be different.

**Comment by proneon267** [comment]
I also noticed another bug while testing: windows can have the same IDs, which defeats the purpose of ensuring uniqueness. Additionally, I didn't find any apparent way to retrieve a window by its ID, other than by iterating through all windows with the following code:
```
for window in app.windows:
    if window.id == CHOSEN_WINDOW_ID:
        # Found the window
        break
```
Is this the correct approach, or am I missing something?
If not, could we make the Window Registry API symmetrical to the Widget Registry? This would ensure that only windows with unique IDs are added to the windows registry, and also allow us to retrieve a window by its ID using:
```
app.windows[CHOSEN_WINDOW_ID]
```

**Comment by freakboy3742** [comment]
>.. could we make the Window Registry API symmetrical to the Widget Registry? This would ensure that only windows with unique IDs are added to the windows registry, and also allow us to retrieve a window by its ID using:

I can't see any reason we shouldn't allow lookup-by-ID for windows; in fact, most (if not all) of the implementation of WidgetRegistry should be reusable as well. 

However, that's new feature, not a bug, and it should be handled independently of this PR.

File: core/tests/test_window.py [hunk 5]
```diff
@@ -381,6 +381,72 @@ def test_screen_position(window, app):
...
+    LABEL_WIDGET_ID = "sample_label"
+
+    label_widget = toga.Label(text="Sample Label", id=LABEL_WIDGET_ID)
+    second_window_content = toga.Box(id=CONTENT_WIDGET_ID, children=[label_widget])
+    # Creating widgets with same ID is allowed but it should
+    # raise KeyError only when assigned to a window.
+    try:
+        third_window_content = toga.Box(id=CONTENT_WIDGET_ID)
+    except KeyError:
+        raise AssertionError(
```
Comment (freakboy3742) [inline core/tests/test_window.py:398]: This would usually be `pytest.fail()`, not a manual raise of AssertionError.

File: core/tests/test_window.py [hunk 6]
```diff
@@ -381,6 +381,72 @@ def test_screen_position(window, app):
...
+    second_window.content = second_window_content
+    assert CONTENT_WIDGET_ID in app.widgets
+    assert LABEL_WIDGET_ID in app.widgets
+    # Assigning widget with same widget ID to be a window's content should raise a KeyError
+    try:
+        third_window.content = third_window_content
+    except KeyError:
+        assert True
+    else:
+        raise AssertionError(
```
Comment (freakboy3742) [inline core/tests/test_window.py:419]: This whole block should be `with pytest.raises`.

File: core/tests/test_window.py [hunk 7]
```diff
@@ -381,6 +381,72 @@ def test_screen_position(window, app):
...
+            "Assigning a widget with same ID to be a window's content should have raised KeyError."
+        )
+    assert CONTENT_WIDGET_ID not in third_window.widgets
+
+    # Creating a new widget with same widget ID should not raise KeyError
+    try:
+        new_label_widget = toga.Label(text="Sample Label", id=LABEL_WIDGET_ID)
+    except KeyError:
+        raise AssertionError(
```
Comment (freakboy3742) [inline core/tests/test_window.py:428]: As above, this is `pytest.fail()`

File: core/tests/test_window.py [hunk 8]
```diff
@@ -381,6 +381,72 @@ def test_screen_position(window, app):
...
+            "Reusing same ID to create another widget when not assigned to a window "
+            "should not have raised KeyError."
+        )
+    # But adding the widget to another window's content should raise KeyError
+    try:
+        third_window.content = toga.Box(children=[new_label_widget])
+    except KeyError:
+        assert True
+    else:
+        raise AssertionError(
```
Comment (freakboy3742) [inline core/tests/test_window.py:438]: As above, this is `with pytest.raises`

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
A couple of corrections on how you're using (or not using) pytest test mechanisms.

**Comment by proneon267** [comment]
Thanks for the tips. The `with pytest.raises` is less verbose than `try` `except ` `else` blocks, but I didn't notice any way to specify the custom error message with `with pytest.raises`. Is that the intended behavior or am I missing something?

**Comment by freakboy3742** [comment]
> Thanks for the tips. The `with pytest.raises` is less verbose than `try` `except ` `else` blocks, but I didn't notice any way to specify the custom error message with `with pytest.raises`. Is that the intended behavior or am I missing something?

```
with pytest.raises(ValueError, match=r"You forgot to include .* in your arguments"`:
    ...
```

The `match` argument is a regular expression that will be compared against the message of the exception raised. There's no need to customise the message of the *test* - the default return value is clear enough.

**Comment by proneon267** [comment]
Thanks, I have changed the test to use pytest helpers. Let me know if any other changes are needed. I am currently working on #2484 and will start a new PR for changing the window registry implementation when #2484 will near completion.

**Review by freakboy3742 — APPROVED** [review]
I've made a couple more tweaks to the test to protect against some remaining edge cases, and make the exception tests a little more robust; but otherwise this looks good. Thanks for the fix!

While I don't want to dampen your enthusiasm, I would be wary of having *too* many open PRs at once. We definitely appreciate both bug fixes and feature improvements, but the PRs you're currently producing consistently require multiple passes and improvements before they get to a final form. We're happy to provide the feedback needed to land these PRs; but when you have multiple PRs in flight at once, it means we (the maintainers/reviewers) end up spending a lot of time providing the same sort of feedback on multiple PRs, rather than having getting one PR to completion, then incorporating the experience you've gained in landing one PR into making the next PR better. 

By my count, you currently have 7 open PRs; I'd vastly prefer to see these land before you start opening new PRs - *especially* if you're proposing adding a feature.

**Comment by proneon267** [comment]
I agree with you. I will complete the remaining PRs first and during this time, I will only open a new PR if any of them will be necessary for landing the existing open PRs. Thank you helping as always 😄
